### ISSUE TITLE
IDE0200 recommends change that increases allocation

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#73147]
**Version Used**: 

**Steps to Reproduce**:
```C#
public class C 
{
    private static readonly Invoker s_invoker = new();
    
    public void M1() => Use(s_invoker.Invoke);
    public void M2() => Use(text => s_invoker.Invoke(text));
    
    public void Use(Func<string, int> func) {}
}

public class Invoker
{
    public int Invoke(string text) => 42;
}
```

**Diagnostic Id**:
IDE0200

**Expected Behavior**:
No diagnostic, or at least a different diagnostic code indicating the change will be making the code less efficient.

**Actual Behavior**:
IDE0200 gets raised on M2 and recommends it be changed to the equivalent of M1. But M1 always allocates a new delegate whereas M2 as written above caches the delegate such that ammortized it only ever allocates once.

**Comment by CyrusNajmabadi** [comment]
Our view here when we last looked at this was that for code that wants to ensure perf above clarity here, it should mark the lambda explicitly to ensure it has non-alloc semantics.  So, in this case, add `static` to the lambda.  Do that, and we won't touch it.

**Comment by stephentoub** [comment]
"Unnecessary" suggests it serves no purpose. That is of course not the case.

**Comment by sharwell** [comment]
I agree with @stephentoub here. IDE0200 certainly attempts to avoid being reported in cases where a new allocation is introduced. For this case, it appears it was missed _either_ because we did not previously optimize `M2()` or because we missed this edge case in testing.

**Comment by CyrusNajmabadi** [comment]
This case was not missed.  IT was an intentional part of the original design here.  And it's one of the main reasons i championed and implemented teh static-lambda proposal.  The general feeling being that it's fine to have clear code that occasionally allocs, and if you're in a perf sensitive domain, you take the steps necessary to annotate your code to make that clear.

**Comment by stephentoub** [comment]
At a minimum, the "unnecessary" in "Remove unnecessary lambda expression" should be clarified. The lambda or something in its stead are in fact necessary to avoid the allocation. If we can improve the compiler to actually avoid the allocation, all the better.

**Comment by stephentoub** [comment]
> You'll get the same result

You won't, actually: it is observable, in that the delegate you get in one case points directly to the target method and in another case points to a compiler-generated one. e.g. this:
```C#
C c = new();
c.M1();
c.M2();

public class C
{
    private static readonly Invoker s_invoker = new();

    public void M1() => Use(s_invoker.Invoke);
    public void M2() => Use(text => s_invoker.Invoke(text));

    public void Use(Func<string, int> func) => Console.WriteLine(func.Target);
}

public class Invoker
{
    public int Invoke(string text) => 42;
}
```
prints:
```
Invoker
C+<>c
```

I'm not suggesting we remove the analyzer. I'm simply suggesting at a minimum we clarify what we mean. It's not at all obvious, especially for someone doing a bulk fix-all, that this can silently lead to additional allocation happening.

**Comment by CyrusNajmabadi** [comment]
>  I'm simply suggesting at a minimum we clarify what we mean. 

I'm ok with that.
